File: LAB_02/src/KI304/Stefaniv/Lab2/Laces.java
package KI304.Stefaniv.Lab2;

import java.util.Objects;

/**
 * The `Laces` class represents shoe laces with properties such as length and color.
 */
public class Laces {
    private int length;
    private String color;

    /**
     * Constructor for initializing a laces object with specified properties.
     *
     * @param length The length of the laces.
     * @param color  The color of the laces.
     * @throws IllegalArgumentException Thrown if the length value is less than or equal to zero.
     */
    public Laces(int length, String color) {
        if(length <= 0) {
            throw new IllegalArgumentException("Laces cannot be less than or equal to 0");
        }
        this.length = length;
        this.color = color;
    }

    /**
     * Method for getting the length of the laces.
     *
     * @return The length of the laces.
     */
    public int getLength() {
        return length;
    }

    /**
     * Method for setting the length of the laces.
     *
     * @param length The new length of the laces.
     * @throws IllegalArgumentException Thrown if the length value is less than or equal to zero.
     */
    public void setLength(int length) {
        if(length <= 0) {
            throw new IllegalArgumentException("Laces cannot be less than or equal to 0");
        }
        this.length = length;
    }

    /**
     * Method for getting the color of the laces.
     *
     * @return The color of the laces.
     */
    public String getColor() {
        return color;
    }

    /**
     * Method for setting the color of the laces.
     *
     * @param color The new color of the laces.
     */
    public void setColor(String color) {
        this.color = color;
    }

    /**
     * Overridden method for getting the string representation of the laces.
     *
     * @return The string representation of the laces.
     */
    @Override
    public String toString() {
        return "Length: " + length + ", " + "color: " + color;
    }

    /**
     * Overridden method for comparing laces objects.
     *
     * @param o The object to compare the laces with.
     * @return true if the laces are identical to the specified object, otherwise false.
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Laces laces)) return false;
        return length == laces.length && Objects.equals(color, laces.color);
    }

    /**
     * Overridden method for calculating the hash code of the laces.
     *
     * @return The hash code of the laces.
     */
    @Override
    public int hashCode() {
        return Objects.hash(length, color);
    }
}

File: LAB_09/laces.py
class Laces:
    # Initialize the Laces object with a given length and color.
    def __init__(self, length, color):
        # Check if the length is less than or equal to 0, and raise a ValueError if so.
        if length <= 0:
            raise ValueError("Laces cannot be less than or equal to 0")
        # Set the length and color attributes of the Laces object.
        self.length = length
        self.color = color

    # Getter method to retrieve the length of the laces.
    def get_length(self):
        return self.length

    # Setter method to update the length of the laces, with validation to ensure it's not less than or equal to 0.
    def set_length(self, length):
        if length <= 0:
            raise ValueError("Laces cannot be less than or equal to 0")
        self.length = length

    # Getter method to retrieve the color of the laces.
    def get_color(self):
        return self.color

    # Setter method to update the color of the laces.
    def set_color(self, color):
        self.color = color

    # Custom string representation of the Laces object.
    def __str__(self):
        return f"Length: {self.length}, Color: {self.color}"

    # Custom equality comparison method for Laces objects.
    def __eq__(self, other):
        # Check if the objects are identical in memory.
        if self is other:
            return True
        # Check if the other object is an instance of Laces, and compare length and color attributes.
        if not isinstance(other, Laces):
            return False
        return self.length == other.length and self.color == other.color

    # Custom hash method to allow Laces objects to be used in sets and dictionaries.
    def __hash__(self):
        # Use a tuple of length and color to generate a hash value.
        return hash((self.length, self.color))
